plutus-core/testlib/PlutusCore/Generators/PIR/Builtin.hs

```diff
@@ -0,0 +1,144 @@
+-- We don't have higher-kinded built-in types, so we don't do this kind of shrinking for any kinds
+-- other than *.
+shrinkDefaultUniApplyStar :: DefaultUni (Esc b) -> [MaybeSomeTypeOf GHC.Type]
+shrinkDefaultUniApplyStar (fun `DefaultUniApply` arg) =
+    [JustSomeType arg | SingType <- [toSingKind arg]] ++ shrinkDefaultUniApplyStar fun
+shrinkDefaultUniApplyStar _ = []
+
+instance KnownKind k => Arbitrary (MaybeSomeTypeOf k) where
+   arbitrary = do
+       size <- getSize
+       oneof $ case knownKind @k of
+           SingType ->
+               [genDefaultUniApply | size > 10] ++
+               [ pure $ JustSomeType DefaultUniInteger
+               , pure $ JustSomeType DefaultUniByteString
+               , pure $ JustSomeType DefaultUniString
+               , pure $ JustSomeType DefaultUniUnit
+               , pure $ JustSomeType DefaultUniBool
+               -- Commented out, because the 'Arbitrary' instance for 'Data' isn't implemented
+               -- (errors out).
+               -- , pure $ JustSomeType DefaultUniData
+               ]
+           SingType `SingKindArrow` SingType ->
+               [genDefaultUniApply | size > 10] ++
+               [pure $ JustSomeType DefaultUniProtoList]
+           SingType `SingKindArrow` SingType `SingKindArrow` SingType ->
+               -- No 'genDefaultUniApply', because we don't have any built-in type constructors
+               -- taking three or more arguments.
+               [pure $ JustSomeType DefaultUniProtoPair]
+           _ -> [pure NothingSomeType]
+
+   -- No shrinks if you don't have anything to shrink.
+   shrink NothingSomeType    = []
+   shrink (JustSomeType uni) =
+       case knownKind @k of
+           SingType -> case uni of
+               -- 'DefaultUniUnit' is the "minimal" built-in type, can't shrink it any further.
+               DefaultUniUnit -> []
+               -- Any other built-in type of kind * shrinks to 'DefaultUniUnit' and if it happens to
+               -- be a built-in type application, then also all suitable arguments of the
+               -- application.
+               _              -> JustSomeType DefaultUniUnit : shrinkDefaultUniApplyStar uni
+           -- No suitable builtins to shrink to for non-* kinds.
+           _ -> []
+
+-- | Convert a @MaybeSomeTypeOf kind@ to a @Maybe (SomeTypeIn DefaultUni)@ and forget the @kind@
+-- (which was only used for generating well-kinded built-in type applications, which we have to do
+-- due to having intrinsic kinding of builtins).
+fromMaybeSomeTypeOf :: MaybeSomeTypeOf kind -> Maybe (SomeTypeIn DefaultUni)
+fromMaybeSomeTypeOf NothingSomeType    = Nothing
+fromMaybeSomeTypeOf (JustSomeType uni) = Just $ SomeTypeIn uni
+
+-- | Generate a built-in type of a given kind.
+genBuiltinTypeOf :: Kind () -> Gen (Maybe (SomeTypeIn DefaultUni))
+genBuiltinTypeOf kind =
+    withKnownKind kind $ \(_ :: Proxy kind) ->
+        fromMaybeSomeTypeOf <$> arbitrary @(MaybeSomeTypeOf kind)
```

Comment:
:exploding_head:

Reply:
No, it's cool!

Reply:
Now that there's a chance we can get away with not having any of this intrinsic typing by having a type-level and a term-level universe, I'm starting to hate all this likely unnecessary complexity.